oidc/oidc.go

```diff
@@ -201,35 +203,53 @@ func (p *ProviderConfig) NewProvider(ctx context.Context) *Provider {
 	}
 }
 
+var oidcConfigCache, _ = ristretto.NewCache(&ristretto.Config[string, []byte]{
+	NumCounters: 10_000,
+	MaxCost:     10_000 * 1024 * 1024, // 10k items, 1MB each
+	BufferItems: 64,
+})
+
 // NewProvider uses the OpenID Connect discovery mechanism to construct a Provider.
 //
 // The issuer is the URL identifier for the service. For example: "https://accounts.google.com"
 // or "https://login.salesforce.com".
 func NewProvider(ctx context.Context, issuer string) (*Provider, error) {
 	wellKnown := strings.TrimSuffix(issuer, "/") + "/.well-known/openid-configuration"
-	req, err := http.NewRequest("GET", wellKnown, nil)
-	if err != nil {
-		return nil, err
-	}
-	resp, err := doRequest(ctx, req)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
 
-	body, err := io.ReadAll(resp.Body)
-	if err != nil {
-		return nil, fmt.Errorf("unable to read response body: %v", err)
-	}
+	var p providerJSON
+	var body []byte
 
-	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("%s: %s", resp.Status, body)
-	}
+	if cached, found := oidcConfigCache.Get(wellKnown); found {
+		err := json.Unmarshal(cached, &p)
+		if err != nil {
+			return nil, fmt.Errorf("oidc: failed to decode provider discovery object: %v", err)
+		}
+		body = cached
+	} else {
+		req, err := http.NewRequest("GET", wellKnown, nil)
+		if err != nil {
+			return nil, err
+		}
+		resp, err := doRequest(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		defer resp.Body.Close()
 
-	var p providerJSON
-	err = unmarshalResp(resp, body, &p)
-	if err != nil {
-		return nil, fmt.Errorf("oidc: failed to decode provider discovery object: %v", err)
+		body, err = io.ReadAll(io.LimitReader(resp.Body, 1<<20))
+		if err != nil {
+			return nil, fmt.Errorf("unable to read response body: %v", err)
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("%s: %s", resp.Status, body)
+		}
+		oidcConfigCache.Set(wellKnown, body, int64(len(body)))
```

Comment:
Consider using gzip to compress the cache if you expect the cache to be large (which I don't think it will be)